Fix adjustedPath on Windows

juliaberkhout opened the pull request:
Commit 4293b548379d1ebe3728d6b20d10d8e5cd97c51f removes the code that replaces the File.seperator with a forward slash.
This replacement is necessary on Windows, so this removal makes that resources from a JAR can no longer be found while trying to add them to the archive.

Example: See https://github.com/arquillian/arquillian-core/blob/master/protocols/servlet/src/main/java/org/jboss/arquillian/protocol/servlet/v_3/ProtocolDeploymentAppender.java#L41. The resource "org/jboss/arquillian/protocol/servlet/v_3/web-fragment.xml" is added, which leads to this method https://github.com/shrinkwrap/shrinkwrap/blob/main/impl-base/src/main/java/org/jboss/shrinkwrap/impl/base/container/ContainerBase.java#L805. On Windows, `adjustedPath` = org\jboss\arquillian\protocol\servlet\v_3\web-fragment.xml and the resource cannot be found, which leads to the IllegalArgumentException thrown on L824. Replacing `adjustedPath` with org/jboss/arquillian/protocol/servlet/v_3/web-fragment.xml solves this problem.

juliaberkhout commented on impl-base/src/main/java/org/jboss/shrinkwrap/impl/base/container/ContainerBase.java:
```diff
@@ -811,7 +811,7 @@ private T addNestedJarFileResource(final File resource, final ArchivePath target
         final int indexOfExclamationPoint = path.indexOf('!' + File.separator);
         if(indexOfExclamationPoint!=-1){
-            adjustedPath = path.substring(indexOfExclamationPoint + 2, path.length());
+            adjustedPath = path.substring(indexOfExclamationPoint + 2, path.length()).replace(File.separator, "/");
```
I put the replace of the separator inside the if-statement. Because this method adds a resource from a JAR, isn't indexOfExclamationPoint always != -1?

ALRubinger commented:
Thank you @juliaberkhout!

Is there a failing test in the ShrinkWrap suite on Windows? If not, can we add one which isolates the problem and shows the failure to this PR?

juliaberkhout commented:
I have in fact multiple test failures on Windows (with and without my fix) so maybe there are more bugs like this one. I'll look into that!

juliaberkhout commented:
@ALRubinger There was one failing test, `DynamicContainerTestBase#testAddManifestStringTargetResourceFromJar`, which is exactly the situation I fixed with this PR. But in this testcase, `indexOfExclamationPoint` is actually -1, so that answers my own question about where to put the `replace` (not _in_ the if-statement). With that change, there are no more test failures. I fixed this with a `git commit --amend`.

ALRubinger commented:
Appreciate that!

I'm prepping a Windows install to review the PR and get this upstream.

juliaberkhout commented:
Hi @ALRubinger, are there any updates on the review of this PR?

bartoszmajsak commented:
@ALRubinger @juliaberkhout it seems to be working fine. I created #148 to verify that. This PR should be merged first, as mine included this fix -without it we already have a bunch of tests failing.

juliaberkhout commented:
@bartoszmajsak I'm not sure what you mean (and what should be merged first). All is fine by me, in the end I want to use a version of shrinkwrap that includes this fix.    

So: do you want me to merge this PR? Or should you guys do something first?

bartoszmajsak commented:
@juliaberkhout What I meant is that my PR (#148) has Windows CI job and your fix cherry-picked. I'm suggesting is to merge this PR first to have the fix and proper attribution preserved and only then bring Windows job by squashing commits in my #148 PR.

I leave the merge and review up to @ALRubinger.

juliaberkhout commented:
@ALRubinger Any updates? We would really appreciate the merge of this fix, because we cannot run our Arquillian tests on Windows at this moment.

robinvandenbogaard commented:
Any ETA on when this fix can be applied?

bartoszmajsak commented:
Pinging @ALRubinger

petrberan reviewed (approved):
Sorry it took so long @juliaberkhout . I'm approving and merging this PR, I've tested this on my fork and the CI on Windows passes with this fix.

I'll release a new beta with the fix soon, we have some trouble syncing our Nexus with Maven central, so I suspect next week
